Allow Symfony uid as primary key for Entities managed with Sonata

jordisala1991 opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 4.x is for everything backwards compatible, like patches, features and deprecation notices
    - 5.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/4.x/CONTRIBUTING.md#base-branch
-->
I am targeting this branch, because this is BC.

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

Closes https://github.com/sonata-project/SonataAdminBundle/issues/7327.

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: https://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataDoctrineORMAdminBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS
    - Don't repeat the verb in the messages (don't use "Added", "Changed", etc).
    - Don't use a full stop at the end of the messages.
-->
```markdown
### Added
- Support for `symfony/uid` as primary keys using `Uuid` or `Ulid`

### Fixed
- `ModelFilter` for related entities using compound ids
- `ModelFilter` when filtering for not equals and using an inverse side relation
```

<!--
    If this is a work in progress, uncomment the "To do" section.
    You can add as many tasks as you want.
    If some are not relevant, just remove them.
-->
<!--
## To do

- [ ] Update the tests;
- [ ] Update the documentation;
- [ ] Add an upgrade note.
-->

jordisala1991 committed:
Fix filter, batch delete

jordisala1991 commented:
This is still WIP @VincentLanglet , but I wanted to get your opinion and also see if builds passes or not. Currently I saw the following problems with Uuid and Ulids:

* Batch delete when selecting item by item does not work ✅ 
* Filter when using uuids as a primary key filter (string) does not work ✅ 
* Filter when the uuid is on a foreign key does not work (ModelTypeFilter) ✅ 
* Filter when the uuid is on a foreign key does not work (ModelAutoCompleteFilter) 🔴 

Last one I am fixing right now. Besides that, everything else seems to work...

the thing with uuid is that we can't rely on the __toString method to make the query to the database. We have to rely on the third parameter of the setParameter to tell doctrine which type it is, so it can properly deal with the value. At the same time we have to keep the __toString conversion on the modelManager, since when we print Uuids to the end user, we should print the normal string representation, which might not be the same as is stored (binary in some cases)

Also, I have made 2 more changes:
1. We have some other tests for Uuids (not the Uuid from Symfony, but it seems like a generic implementation. I have moved those fake types to another name. (I guess those tests are meant to be used with an implementation similar to ramsey/uuid)
2. On doctrine bridge 4.4 there are no types for uuid and ulid... so I have to copy that code for testing purposes.

VincentLanglet commented:
Seems nice

VincentLanglet commented:
You say it closes https://github.com/sonata-project/SonataAdminBundle/issues/7327 but is there nothing to do on the Mongo side ?

jordisala1991 commented:
> You say it closes [sonata-project/SonataAdminBundle#7327](https://github.com/sonata-project/SonataAdminBundle/issues/7327) but is there nothing to do on the Mongo side ?

TBH I am not sure how it should be implemented on Mongo side... not even sure if Symfony team or Doctrine team has added UuidTypes, similar to how it is done on Doctrine bridge for the ORM.

IMHO if the issue ends up being persistence related only. After this PR gets merged, we can move the issue to MongoDB persistence to see if there is interest on this implementation.

jordisala1991 commented on src/Filter/ModelFilter.php:
```diff
@@ -140,4 +146,29 @@ private function getParentAlias(ProxyQueryInterface $query, string $alias): stri
+    /**
+     * @param ProxyQueryInterface<object> $query
+     */
+    private function transformSearchValues(ProxyQueryInterface $query, array $values): array
```
This implementation is a bit problematic, but it is needed since: https://github.com/ramsey/uuid-doctrine/issues/18

I know if it is not symfony/uid, but it is happening the same, doctrine has a limitation when passing an array of object to the setParameter. It tries to convert everything to the __toString representation and can't be changed.

Here I tried to iterate over those values, extract the identifier, extract the type of that identifier (assuming the primary key is only one field...) and try to convert to the database value needed to search.

It works for single PK entities, not sure what would happen for compound primary keys. Also not sure if it is something that can happen. This filter is used on an entity with a foreign key to another (that foreign key is only allowed to be on a single field, right?

wdyt @VincentLanglet ?

jordisala1991 replied:
Something that I might read later: https://www.doctrine-project.org/projects/doctrine-orm/en/2.14/tutorials/composite-primary-keys.html

jordisala1991 replied:
Another option to try mitigate the problem could be unroll the loop into multiple andWheres one for each entity in the list. Even in that case, the problem will be the same for each item. I will need to know its identifier, its type, and transform it to the database representation, because I can't pass the full object.

jordisala1991 replied:
I have refactored the code to be simpler but handle way more cases.

jordisala1991 commented on src/Filter/ModelFilter.php:
```diff
@@ -140,4 +139,37 @@ private function getParentAlias(ProxyQueryInterface $query, string $alias): stri
         return $parentAlias;
     }
+
+    private function buildMultipleAnd(ProxyQueryInterface $query, string $alias, FilterData $data): Orx
```
This function will now do the magic....

For each entity that is added to the filter, it finds its identifiers, and for each identifier it adds a condition to the query, if the identifier value is an object and the type is known, is added to the setParameter also.

This works in all scenarios.

* Single identifiers as integers
* Single identifiers as object (uuid)
* Compound identifiers

jordisala1991 replied:
I need to add tests for this. For now I have tested it with the entities provided in the test folder.

jordisala1991 replied:
What I like less is having to replicate this same logic for the autocomplete type, wdyt @VincentLanglet ?

VincentLanglet replied:
The ModelAutocomplete Filter is deprecated

jordisala1991 replied:
Then we can simply not to the change there. I will add a test for the autocomplete Type using the modelFilter, so we are covered.

VincentLanglet commented on tests/App/AppKernel.php:
```diff
@@ -107,6 +110,17 @@ protected function configureContainer(ContainerBuilder $container, LoaderInterfa
         $loader->load(__DIR__.'/config/services.php');
+
+        /*
+         * TODO: Remove all this and related Types when support for Symfony 4.4 is dropped.
```
We could drop Sf 4.4 now no ?

jordisala1991 replied:
I would prefer to do it on another PR. Also I would like to do it for all bundles at once.

jordisala1991 commented:
I will need to also take a look at why test for another unrelated feature are failing now.

jordisala1991 commented on src/Model/ModelManager.php:
```diff
@@ -336,9 +337,10 @@ public function addIdentifiersToQuery(string $class, BaseProxyQueryInterface $qu
                 $parameterName = sprintf('field_%s_%s_%d', $prefix, $name, $pos);
                 $ands[] = sprintf('%s.%s = :%s', current($qb->getRootAliases()), $name, $parameterName);
-                $qb->setParameter($parameterName, $ids[$posName]);
+                $qb->setParameter($parameterName, $ids[$posName], $fieldType);
```
This will be a bit problematic if the id is for example a DateTime... We need to find a better solution for this. Will also add tests for it.

jordisala1991 replied:
Just tested, since doctrine does not support DateTime as a primary key it is not a problem for that kind of field.

jordisala1991 commented on composer.json:
```diff
@@ -36,7 +36,7 @@
         "symfony/console": "^4.4 || ^5.4 || ^6.2",
         "symfony/dependency-injection": "^4.4 || ^5.4 || ^6.2",
-        "symfony/doctrine-bridge": "^4.4 || ^5.4 || ^6.2",
+        "symfony/doctrine-bridge": "^4.4 <4.4.42 || ^5.4 || ^6.2",
```
From 4.4.42 my tests no longer works (it works for 5.4 and 6.2):

there are 3 PRs that affect this release:
https://github.com/symfony/symfony/pull/46443
https://github.com/symfony/symfony/pull/46098
https://github.com/symfony/symfony/pull/46282

I will try to investigate which one is make tests failing... For now, what I see is that with 4.4.42 I get an empty ArrayCollection directly from the EntityType.

jordisala1991 replied:
https://github.com/symfony/symfony/pull/46282 This one is making test fail.

If I manually revert on vendor code, it starts working. Maybe this condition depends on the ORM version? ping @derrabus

jordisala1991 replied:
It is strange that it works with prefer-lowest, 5.4, 6.2, but not with 4.4 (higher deps but fixed to 4.4). Will debug more to see if it is fixable on our side, but any help is welcome.

derrabus replied:
I don't know if I can be of any help here. I'm pretty sure that the change you've linked is correct.

That being said, Symfony 4 is dead. Delivering features for an EOL Symfony version does not seem like a sane thing to do. Hunting down broken tests for an EOL Symfony version seems like a waste of time to me.

VincentLanglet replied:
I agree. That's why I suggested to drop 4.4 @jordisala1991. You can drop 4.4 first in another PR if you want.

jordisala1991 replied:
Solved.

jordisala1991 committed:
Add tests for Uuids and Ulids as primary keys

jordisala1991 commented on tests/Functional/Admin/BaseAdminTestCase.php:
```diff
@@ -0,0 +1,132 @@
+
+use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
+
+abstract class BaseAdminTestCase extends WebTestCase
```
WDYT about doing this class on src and marking it as experimental? This way we can reuse it on other bundles while not worrying for now about BC promises (we can remove experimental later once the class is stable enough) @VincentLanglet

VincentLanglet replied:
You mean doing this on the src class of SonataAdmin (instead of DoctrineORM) ?

It makes me think of this https://github.com/sonata-project/SonataAdminBundle/issues/7648

jordisala1991 replied:
Yes, tbh this class is quite simple. Not near as complex as the issue you mention. Also I think that proposal was more about adding some asserts for Sonata things (also a nice proposal). This one is like an abstraction around tipical things that you can test on an admin (Crud, filters, batch actions).

phansys replied:
See https://github.com/sonata-project/SonataCoreBundle/tree/3.x/src/CoreBundle/Test.

jordisala1991 replied:
Then, should we expose this as experimental? Or we can do it later on another PR if needed?

jordisala1991 replied:
Lets for now not expose this class.

jordisala1991 commented on src/Filter/ModelFilter.php:
```diff
@@ -76,31 +77,20 @@ protected function handleMultiple(ProxyQueryInterface $query, string $alias, Fil
-                );
-            } else {
-                $or->add($query->getQueryBuilder()->expr()->isNull(
-                    sprintf('IDENTITY(%s.%s)', $this->getParentAlias($query, $alias), $this->getFieldName())
```
We didn't had enough tests when this was added. It does not work for inverse relations. For example the UlidChildEntityAdmin filtering on the parent field. It is not related to the uuid, it simply does not work AFAIK.

@VincentLanglet do you found something like that on your previous projects?

jordisala1991 replied:
The error only happens when someone is trying to filter with the advanced filter "NOT EQUALS"

VincentLanglet replied:
I'm not sure I worked with the ModelFilter. I think I always use ChoiceFilter by selecting the id field.

jordisala1991 replied:
I think I found a solution, let's see

jordisala1991 commented on src/Filter/ModelFilter.php:
```diff
@@ -76,31 +77,24 @@ protected function handleMultiple(ProxyQueryInterface $query, string $alias, Fil
-                    sprintf('%s.%s IS EMPTY', $this->getParentAlias($query, $alias), $this->getFieldName())
-                );
+            if (false === $this->getAssociationMapping()['isOwningSide']) {
+                $nullExpression = sprintf('IDENTITY(%s.%s) IS NULL', $alias, $this->getAssociationMapping()['mappedBy']);
```
wdyt @VincentLanglet ?

jordisala1991 commented:
The changelog could be improved, IMO I also fixed:

* ModelFilter with compound entity
* ModelFilter on the inverse side of the relation (when filtering for not equals)

Edit: Added.
